containers.go

```diff
@@ -696,6 +699,26 @@ func (r *containerStore) Create(id string, names []string, image, layer, metadat
 		GIDMap:         copyIDMap(options.GIDMap),
 		volatileStore:  options.Volatile,
 	}
+	defer func() {
+		if err != nil {
+			// undo all that we might have done after this point
+			if e := os.RemoveAll(r.datadir(container.ID)); e != nil && !errors.Is(e, os.ErrNotExist) {
+				logrus.Debugf("while cleaning up data directory for container %q we failed to create: %v", id, e)
+			}
+			for _, name := range names {
+				delete(r.byname, name)
+			}
+			delete(r.bylayer, layer)
+			r.idindex.Delete(id)
+			delete(r.byid, id)
+			r.containers = containerSliceWithoutValue(r.containers, container)
```

Comment:
Hmm, I like using `Delete()` as a cleanup.  Switching to doing that, and to set `Flags` in the just-created container structure rather than modifying a passed-in `options.Flags` map and then copying the whole map into the just-created container structure.

In order to correctly clean up indexes that it has already modified, it looks like `layerStore.PutAdditionalLayer()` should have been using `Delete()` for cleanup as well, changing it.